### DISCUSSION TITLE
Custom format for assertion exception messages

### DISCUSSION THREAD

**Discussion by eylvisaker** [fluentassertions/fluentassertions#2218] [Q&A]
Is there a way to customize the format of the exception messages? In particular I would like to move the "because" clause to the end. 
I get a lot of failed tests with messages that look like this:

* Expected serverResponse.IsSuccess to be true because The Name field is required., but found False.

but I feel they would read much more naturally like this:

* Expected serverResponse.IsSuccess to be true but found False, because The Name field is required.

Sometimes it's really hard to craft good, concise "should succeed because" clause to explain why an assertion is true, but it's very easy to to automatically have a concise, helpful message that "but failed because" clause. A format with the "because" at the end would read well in either case.

**Comment by dennisdoomen** [comment]
No, unfortunately not. The "because" part is part of a hard-coded template that is specific for every assertion.

**Comment by djrhodes** [comment]
I agree with the OP, sometimes it can be useful  to provide some context around the failure to assist the dev, but it doesn't make sense to provide it in the `because` argument due to where it's integrated in the failure message. As far as I can see, the options available to provide this context when using FA is either to catch the assertion exception and write the context to stdout : 
```
try
{
    result.IsSuccessful.Should().BeTrue();
}
catch
{
    TestContext.Out.WriteLine($"Custom failure message with {result.Context}");
    throw;
}
```
or use FailWith : 
```
if (!result.IsSuccessful)
{
    Execute.Assertion.FailWith($"Custom failure message with {result.Context}");
}
```

With `try... catch` you get both the FA message and the context, with `FailWith` you just get the context message. I went with FailWith since I wasn't using the `because` parameter so I could do without the default FA message. 

Neither are particularly neat when used to working with the fluent experience ;-) I feel that this could be a lot neater if it was handled in the API e.g. could you add an additional optional param - failedBecause : `result.IsSuccesful.Should().BeTrue(failedBecause: $"Custom failure message with {result.Context}")`

**Reply by dennisdoomen** [reply]
Another way to provide context is to wrap the assertion in a `using new AssertionScope("some context")`;

**Reply by whiterose-it** [reply]
> Another way to provide context is to wrap the assertion in a `using new AssertionScope("some context")`;

Thanks for the additional tip, and for the work you do on FA. It really is an awesome library and one of my core libs I use on every project.
